core-services/src/main/java/org/zalando/nakadi/service/publishing/EventPublisher.java

```diff
@@ -211,14 +228,53 @@ private void setEventKey(final List<BatchItem> batch, final EventType eventType)
                     partitionKeyField = DATA_PATH_PREFIX + partitionKeyField;
                 }
                 for (final BatchItem item : batch) {
-                    final JsonPathAccess jsonPath = new JsonPathAccess(item.getEvent());
+                    final JsonPathAccess jsonPath = new JsonPathAccess(item.getEvent().getEventJson());
                     final String eventKey = jsonPath.get(partitionKeyField).toString();
-                    item.setEventKey(eventKey);
+                    item.setKey(eventKey);
                 }
             }
         }
     }
 
+    private void setHeaders(final List<BatchItem> batch, final EventType eventType) {
```

Comment:
Clearly this is wrong place for the method

Reply:
On second thought, I see that, these should be set closer to publishing when `SUBMITTED`. I'll try and fix this. Thanks for pointing out! :)